Browser extension: fix GitHub file info detection

chrismwendt opened the pull request:
Prior to this change, the revision was found by inspecting the branch/tree button and whether or not the text was truncated. That seemed error-prone to me.

After this change, the file path is found first (by inspecting the "Copy path" button), then the revision is derived from that and the URL. This seems a bit more robust and simpler.

Test plan: make sure the "Open on Sourcegraph" button appears in the **file header** for these:

- https://github.com/sourcegraph/sourcegraph/blob/alert-bg/cmd/frontend/internal/session/session.go
- https://github.com/sourcegraph/sourcegraph/blob/01d571ba62853db43d5dd0b73436ec47eca7662b/cmd/frontend/internal/session/session.go

Fixes https://github.com/sourcegraph/sourcegraph/issues/4280

#### Test Plan

##### Code Hosts

- [x] GitHub
- [x] GitHub Enterprise
- [x] Refined GitHub
- ~Phabricator~
- ~Phabricator integration~
- ~Bitbucket~
- ~Gitlab~

##### Browsers

- [x] Chrome
- [x] Firefox

felixfbecker commented on browser/src/libs/github/util.tsx:
```diff
@@ -182,29 +182,30 @@ function getDiffResolvedRevFromPageSource(pageSource: string, isPullRequest: boo
-    }
-    return selectButton.title
+export function getFilePath(): string | undefined {
+    return document.evaluate(
```
I see why XPath is used here but it should be documented why - in general I'd like to avoid it because we can't assume people on the team understand it. Asserting the content would also break if GitHub had different translations (BitBucket does for example, so this shouldn't become standard practice).

Could you benchmark that the performance of this is not worse than the alternative `document.querySelectorAll().find()`?

```suggestion
    // Using XPath here instead of CSS selectors because the only way to know is the content of the button
    // GitHub is always in English
    return document.evaluate(
```

chrismwendt replied:
Switched to CSS `clipboard-copy[value]` to avoid relying on the text being in English and to avoid XPath.

felixfbecker commented on browser/src/libs/github/util.tsx:
```diff
@@ -182,29 +182,30 @@ function getDiffResolvedRevFromPageSource(pageSource: string, isPullRequest: boo
+ *
+ * TODO ideally, this should only scrape the code view itself.
  */
-export function getBranchName(): string {
```
So we don't need this function anywhere else, i.e. don't need to know the branch name anywhere else? o.O

chrismwendt replied:
I was also surprised this `getBranchName` function isn't used anywhere else.

felixfbecker commented:
Please add a test to prevent regressions :)

lguychard commented:
Make sure to test this on GitHub Enterprise as well!

felixfbecker commented on browser/src/libs/github/file_info.ts:
```diff
@@ -80,7 +80,7 @@ export const resolveFileInfo = (codeView: HTMLElement): Observable<FileInfo> =>
 
         const filePath = getFilePath()
-        if (filePath === undefined) {
+        if (!filePath) {
```
Does this mean the function can now return empty strings?

lguychard replied:
Its return type is `string | null`

felixfbecker commented on browser/src/libs/github/file_info.ts:
```diff
@@ -79,23 +79,20 @@ export const resolveFileInfo = (codeView: HTMLElement): Observable<FileInfo> =>
-            commitID,
-            rev: branchName,
+            commitID: getCommitIDFromPermalink(),
+            rev: revAndFilePath.slice(0, revAndFilePath.length - filePath.length),
```
```suggestion
            rev: revAndFilePath.slice(0, -filePath.length),
```

chrismwendt commented:
I switched to `.js-permalink-shortcut` because it seems to be more stable. In GitHub snapshots in our fixtures, either `clipboard-copy` doesn't have the `value` attribute or the first instance of it in the DOM is not the blob path.

I also added some tests.

chrismwendt commented on browser/src/libs/github/util.tsx:
```diff
@@ -182,29 +182,44 @@ function getDiffResolvedRevFromPageSource(pageSource: string, isPullRequest: boo
+        )
+        return null
+    }
+    return path.join('/')
```
I completely changed the implementation of `getFilePath()` here, so needs another review 📝

felixfbecker commented on browser/src/libs/github/util.tsx:
```diff
@@ -182,29 +182,44 @@ function getDiffResolvedRevFromPageSource(pageSource: string, isPullRequest: boo
+    const permalink = document.querySelector('.js-permalink-shortcut')
+    if (!permalink) {
+        console.error(`Unable to determine the file path because no .js-permalink-shortcut element was found.`)
+        return null
```
Imo it's better to throw an Error than to return `null` - an Error should be thrown if a function cannot fulfill its contract, which is the case here. The caller then probably also cannot fulfill _its_ contract, so it's useful to be able to let the exception bubble up (vs a Go style checking for null at every level) and persist the context why it failed vs just a global log message

chrismwendt replied:
Agreed, switched to throwing an error.

felixfbecker commented on browser/src/libs/github/util.tsx:
```diff
@@ -182,29 +182,44 @@ function getDiffResolvedRevFromPageSource(pageSource: string, isPullRequest: boo
+        return null
+    }
+    // @ts-ignore these unused variables aid readability
+    const [_, _user, _repo, _blob, _commitID, ...path] = href.split('/')
```
`@ts-ignore` is _really_ dangerous because it disables _any_ TypeScript error, even type errors. For example, you could change the type of `href` to something other than string here and not get a compile error anymore for calling `split()` on it.

If it's only for documentation, use comments:
```suggestion
    const [, /* user */, /* repo */, /* blob */, /* commitID */, ...path] = href.split('/')
```

felixfbecker replied:
I would also suggest parsing the `href` with `URL` first and only running this logic on `url.pathname` to be resilient against missing protocol `//`, hash fragments (line number?), search queries (unified/split?) etc

chrismwendt replied:
Avoided `@ts-ignore` :+1: Unfortunately, prettier does this:

```typescript
    const [, , , , , /* user */ /* repo */ /* blob */ /* commit ID */ ...path] = href.split('/')
```

So I went with this instead:

```typescript
    // <empty>/<user>/<repo>/blob/<commitID>/<path/to/file>
    const [, , , , , ...path] = href.split('/')
```

chrismwendt replied:
The attribute looks like `href="/user/repo/bob/commitID/path/to/file"`, and that's not a valid URL, so I can't use `URL` to parse it. However, I found out that there's a `.href` attribute on anchor tags with the full URL, so I used that in order to safeguard against GitHub changing the href attribute to a full URL at some point.

felixfbecker replied:
`URL` can parse relative URLs like `new URL(a.href, window.location.href)`

chrismwendt replied:
Neat! I'll stick to `.href` because it's `string`, whereas `getAttribute('href'): string | null`.

felixfbecker commented on browser/src/libs/github/file_info.test.ts:
```diff
@@ -0,0 +1,21 @@
+import { readFile } from 'mz/fs'
+import { getFilePath } from './util'
+
+describe('github/file_info', () => {
```
This is the name of the module, shouldn't there be another describe with the name of the function `resolveDiffFileInfo()`?

chrismwendt replied:
Sure, do you mean `getFilePath()`?

felixfbecker replied:
Yes, sorry. Please wrap in a `describe('getFilePath()', () => { ... })`.

chrismwendt replied:
✅ Wrapped in `describe('getFilePath()', () => { ... })`

chrismwendt commented on browser/src/libs/github/__snapshots__/file_info.test.ts.snap:
```diff
@@ -0,0 +1,9 @@
+
+exports[`github/file_info github.com Refined Github finds the file path 1`] = `"shared/src/api/extension/types/url.ts"`;
+
+exports[`github/file_info github.com Vanilla finds the file path 1`] = `"shared/src/api/extension/types/url.ts"`;
```
Got it, I was wondering if snapshot testing was appropriate here. Thanks for the tips!

I switched away from snapshots to table-driven tests.

felixfbecker commented on browser/src/libs/github/file_info.test.ts:
```diff
@@ -0,0 +1,18 @@
+
+describe('github/file_info', () => {
+    for (const [fixture, expectedFilePath] of tests) {
+        it(`getFilePath() finds the file path in ${fixture}`, async () => {
```
```suggestion
        it(`finds the file path in ${fixture}`, async () => {
```

This should read like an English sentence.

chrismwendt commented:
@felixfbecker Did I address everything?
